src/Codec/Xlsx/Templater.hs

```diff
@@ -105,30 +68,35 @@ fixRowHeights rh r n = insertCopies $ shift removeOriginal
       Just h -> foldr (\x m' -> M.insert x h m') m [r..(r + n -1)]
       Nothing -> m
 
-
-runSheet :: Xlsx -> Int -> (TemplateDataRow, TemplateSettings, [TemplateDataRow]) -> IO Worksheet
-runSheet x n (cdr, ts, d) = do
-  ws <- sheet x n
-  let
-    templateRows = if tsOrientation ts == Columns then transpose $ toList ws else toList ws
-    repeatRow = tsRepeated ts
-    (prolog, templateRow : epilog) = splitAt repeatRow templateRows
-    tpl = buildTemplate repeatRow templateRow
-    prolog' = replacePlaceholders prolog cdr
+applyTemplateOnSheet :: (TemplateDataRow, TemplateSettings, [TemplateDataRow]) -> Worksheet -> Worksheet
+applyTemplateOnSheet (cdr, ts, d) ws =
+    ws & wsColumns           .~ cw
+       & wsRowPropertiesMap  .~ rh
+       & wsCells             %~ applyTemplateOnCellMap
+  where
+    mayTranspose =
+      case tsOrientation ts of
+        Columns -> each . _1 %~ (\(x,y) -> (y,x))
+        _       -> id
+    repeatRow = tsRepeated ts + 1
     n = length d
-    d' = map (applyTemplate tpl) d
-    epilog' = replacePlaceholders epilog cdr
-    output = concat [prolog', d', epilog']
-    result = if tsOrientation ts == Columns then transpose output else output
+    applyTemplateOnCellMap =
+      M.fromList . mayTranspose . concatMap applyTemplateOnCellPos . mayTranspose . M.toList
+    applyTemplateOnCellPos ((row, col), cell)
+      | row < repeatRow = [ ((row,     col), applyTemplateOnCell cdr cell) ]
+      | row > repeatRow = [ ((row + n, col), applyTemplateOnCell cdr cell) ]
+      | otherwise       = [ ((row + i, col), applyTemplateOnCell dc  cell)
+                          | (i, dc) <- zip [0 ..] d ]
     (cw, rh) = if tsOrientation ts == Columns
-                 then (fixColumns (wsColumns ws) (repeatRow + 1) n, wsRowHeights ws)
-                 else (wsColumns ws, fixRowHeights (wsRowHeights ws) (repeatRow + 1) n)
-    in
-   return $ fromList (wsName ws) cw rh result
+                 then (fixColumns (ws ^. wsColumns) repeatRow n, ws ^. wsRowPropertiesMap)
+                 else (ws ^. wsColumns, fixRowHeights (ws ^. wsRowPropertiesMap) repeatRow n)
+
+applyTemplateOnXlsx :: [(TemplateDataRow, TemplateSettings, [TemplateDataRow])] -> Xlsx -> Xlsx
+applyTemplateOnXlsx options x =
+  x & xlSheets .~ zipWith (\(nm,sh) opt -> (nm, applyTemplateOnSheet opt sh)) (x ^. xlSheets) options
```

Comment:
what about using `each . _1 %~` as you've done above? Using 1 lens twice looks suspicious